xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,737 @@
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_eds LB policy.
+ */
+final class EdsLoadBalancer extends LoadBalancer {
+
+  private final Helper helper;
+  private final ChannelLogger channelLogger;
+
+  private final EndpointWatcherImpl.PickerFactory pickerFactory;
+  private final LoadBalancerRegistry lbRegistry;
+  private final ThreadSafeRandom random;
+  private final LoadStatsStore loadStatsStore;
+  private final OrcaPerRequestUtil orcaPerRequestUtil;
+  private final OrcaOobUtil orcaOobUtil;
+
+
+  @Nullable
+  private XdsClient xdsClient;
+  @Nullable
+  private String edsServiceName;
+
+  @CheckForNull
+  private EndpointWatcherImpl endpointWatcher;
+
+  // TODO(zdapeng): Pass in a nullable LoadStatsStore for LRS.
+  EdsLoadBalancer(Helper helper) {
+    this(
+        helper, EndpointWatcherImpl.pickerFactoryImpl,
+        LoadBalancerRegistry.getDefaultRegistry(), ThreadSafeRandom.ThreadSafeRandomImpl.instance,
+        new LoadStatsStoreImpl(), OrcaPerRequestUtil.getInstance(), OrcaOobUtil.getInstance());
+  }
+
+  @VisibleForTesting
+  EdsLoadBalancer(
+      Helper helper,
+      EndpointWatcherImpl.PickerFactory pickerFactory,
+      LoadBalancerRegistry lbRegistry,
+      ThreadSafeRandom random,
+      LoadStatsStore loadStatsStore,
+      OrcaPerRequestUtil orcaPerRequestUtil,
+      OrcaOobUtil orcaOobUtil) {
+    this.helper = checkNotNull(helper, "helper");
+    this.channelLogger = helper.getChannelLogger();
+    // TODO(zdapeng): TBD: handle enableLrs from ClusterUpdate is false/changed.
+    this.loadStatsStore = checkNotNull(loadStatsStore, "loadStatsStore");
+
+    // The following fields are injected for testing
+    this.pickerFactory = checkNotNull(pickerFactory, "pickerFactory");
+    this.lbRegistry = checkNotNull(lbRegistry, "lbRegistry");
+    this.random = checkNotNull(random, "random");
+    this.orcaPerRequestUtil = checkNotNull(orcaPerRequestUtil, "orcaPerRequestUtil");
+    this.orcaOobUtil = checkNotNull(orcaOobUtil, "orcaOobUtil");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+
+    AtomicReference<XdsClient> xdsClientRef = resolvedAddresses.getAttributes()
+            .get(XdsAttributes.XDS_CLIENT_REF);
+    if (xdsClientRef == null) {
+      // TODO(zdapeng): create a new xdsClient if no one exists.
+      throw new UnsupportedOperationException(
+          "XDS_CLIENT_REF attributes not available in resolvedAddresses " + resolvedAddresses);
+    }
+
+    if (xdsClientRef.get() == null) {
+      // The XdsClient has been shutdown. The balancer will soon shutdown.
+      return;
+    }
+
+    if (xdsClient != null) {
+      Preconditions.checkState(
+          xdsClient == xdsClientRef.get(),
+          "The XdsClient is changed");
```

Comment:
Write more descriptive messages, if you think that's something should never happen, highlight it as a bug.